lxd/vm_qemu.go

```diff
@@ -0,0 +1,2329 @@
+		name:         args.Name,
+		node:         args.Node,
+		description:  args.Description,
+		ephemeral:    args.Ephemeral,
+		architecture: args.Architecture,
+		dbType:       args.Type,
+		snapshot:     args.Snapshot,
+		stateful:     args.Stateful,
+		creationDate: args.CreationDate,
+		lastUsedDate: args.LastUsedDate,
+		profiles:     args.Profiles,
+		localConfig:  args.Config,
+		localDevices: args.Devices,
+		expiryDate:   args.ExpiryDate,
+	}
+
+	// Cleanup the zero values.
+	if vm.expiryDate.IsZero() {
+		vm.expiryDate = time.Time{}
+	}
+
+	if vm.creationDate.IsZero() {
+		vm.creationDate = time.Time{}
+	}
+
+	if vm.lastUsedDate.IsZero() {
+		vm.lastUsedDate = time.Time{}
+	}
+
+	ctxMap := log.Ctx{
+		"project":   args.Project,
+		"name":      vm.name,
+		"ephemeral": vm.ephemeral,
+	}
+
+	logger.Info("Creating instance", ctxMap)
+
+	revert := true
+	defer func() {
+		if !revert {
+			return
+		}
+
+		vm.Delete()
+	}()
+
+	// Load the config.
+	err := vm.init()
+	if err != nil {
+		logger.Error("Failed creating instance", ctxMap)
+		return nil, err
+	}
+
+	// Validate expanded config
+	err = containerValidConfig(s.OS, vm.expandedConfig, false, true)
+	if err != nil {
+		logger.Error("Failed creating instance", ctxMap)
+		return nil, err
+	}
+
+	err = containerValidDevices(s, s.Cluster, vm.Name(), vm.expandedDevices, true)
+	if err != nil {
+		logger.Error("Failed creating instance", ctxMap)
+		return nil, errors.Wrap(err, "Invalid devices")
+	}
+
+	// Retrieve the instance's storage pool
+	_, rootDiskDevice, err := shared.GetRootDiskDevice(vm.expandedDevices.CloneNative())
+	if err != nil {
+		return nil, err
+	}
+
+	if rootDiskDevice["pool"] == "" {
+		return nil, fmt.Errorf("The instances's root device is missing the pool property")
+	}
+
+	storagePool := rootDiskDevice["pool"]
+
+	// Get the storage pool ID for the instance.
+	poolID, pool, err := s.Cluster.StoragePoolGet(storagePool)
+	if err != nil {
+		return nil, err
+	}
+
+	// Fill in any default volume config.
+	volumeConfig := map[string]string{}
+	err = storagePools.VolumeFillDefault(storagePool, volumeConfig, pool)
+	if err != nil {
+		return nil, err
+	}
+
+	// Create a new database entry for the instance's storage volume.
+	_, err = s.Cluster.StoragePoolVolumeCreate(args.Project, args.Name, "", db.StoragePoolVolumeTypeVM, false, poolID, volumeConfig)
+	if err != nil {
+		return nil, err
+	}
+
+	if !vm.IsSnapshot() {
+		// Update MAAS.
+		err = vm.maasUpdate(nil)
+		if err != nil {
+			logger.Error("Failed creating instance", ctxMap)
+			return nil, err
+		}
+
+		// Add devices to instance.
+		for k, m := range vm.expandedDevices {
+			err = vm.deviceAdd(k, m)
+			if err != nil && err != device.ErrUnsupportedDevType {
+				return nil, errors.Wrapf(err, "Failed to add device '%s'", k)
+			}
+		}
+	}
+
+	logger.Info("Created instance", ctxMap)
+	// tomp TODO should this event name be changed?
+	vm.state.Events.SendLifecycle(vm.project, "container-created",
```

Comment:
@stgraber shall I update these container references to "virtual-machine-created" and "/1.0/virtual-machines/%s" respectively?